File: crawler.py
import json
import requests


def extract_eids(search_json):
    search_results = search_json.get("search-results", {})
    entries = search_results.get("entry", [])

    if isinstance(entries, dict):
        entries = [entries]

    return [e["eid"] for e in entries if "eid" in e]


def fetch_publication(eid, headers):
    url = f"https://api.elsevier.com/content/abstract/eid/{eid}"
    r = requests.get(
        url,
        headers=headers,
        params={"view": "FULL"},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def ensure_list(value):
    if value is None:
        return []
    if isinstance(value, list):
        return value
    return [value]


def parse_author_name(author):
    if not isinstance(author, dict):
        return None

    indexed_name = author.get("ce:indexed-name")
    if indexed_name:
        return indexed_name

    given_name = author.get("ce:given-name")
    surname = author.get("ce:surname")
    initials = author.get("ce:initials")

    if given_name and surname:
        return f"{given_name} {surname}".strip()
    if surname and initials:
        return f"{surname}, {initials}".strip()
    if surname:
        return surname
    return None


def extract_authors(response):
    authors_block = response.get("authors", {})
    author_entries = ensure_list(authors_block.get("author"))

    authors = []
    for author in author_entries:
        name = parse_author_name(author)
        if name:
            authors.append(name)

    return authors


def extract_publication_fields(pubData):
    response = pubData.get("abstracts-retrieval-response", {})
    core = response.get("coredata", {})

    all_authors = extract_authors(response)
    main_author = all_authors[0] if all_authors else core.get("dc:creator")
    co_authors = all_authors[1:] if len(all_authors) > 1 else []

    return {
        "title": core.get("dc:title"),
        "date_of_publication": core.get("prism:coverDate"),
        "publication_type": core.get("subtypeDescription"),
        "journal": core.get("prism:publicationName"),
        "publisher": core.get("dc:publisher"),
        "volume": core.get("prism:volume"),
        "issue": core.get("prism:issueIdentifier"),
        "pages": core.get("prism:pageRange"),
        "doi": core.get("prism:doi"),
        "citation_count": core.get("citedby-count"),
        "main_author": main_author,
        "co_authors": co_authors,
    }


def print_publication(pub, index):
    print(f"\nPublication #{index}")
    print("-" * 60)
    print(f"Title: {pub.get('title')}")
    print(f"Date of Publication: {pub.get('date_of_publication')}")
    print(f"Type: {pub.get('publication_type')}")
    print(f"Journal: {pub.get('journal')}")
    print(f"Publisher: {pub.get('publisher')}")
    print(f"Volume: {pub.get('volume')}")
    print(f"Issue: {pub.get('issue')}")
    print(f"Pages: {pub.get('pages')}")
    print(f"DOI: {pub.get('doi')}")
    print(f"Citation Count: {pub.get('citation_count')}")
    print(f"Main Author: {pub.get('main_author')}")
    print(f"Co-Authors: {', '.join(pub.get('co_authors', [])) if pub.get('co_authors') else 'None'}")


with open("config.json", "r") as file:
    config = json.load(file)

apikey = config["apikey"]
insttoken = config.get("insttoken")

author_id = "55615806100"

url = "https://api.elsevier.com/content/search/scopus"

headers = {
    "Accept": "application/json",
    "X-ELS-APIKey": apikey,
}

if insttoken:
    headers["X-ELS-Insttoken"] = insttoken

resp = requests.get(
    url,
    headers=headers,
    params={
        "view": "STANDARD",
        "query": f"AU-ID({author_id})",
        "count": 5,
        "start": 0,
    },
    timeout=30,
)
resp.raise_for_status()

search_response = resp.json()
eids = extract_eids(search_response)

if not eids:
    print("No EIDs found.")
    exit()

publications = []

for eid in eids:
    try:
        pubData = fetch_publication(eid, headers)
        limited = extract_publication_fields(pubData)
        publications.append(limited)
    except requests.HTTPError as e:
        print(f"Failed for {eid}: {e}")
    except Exception as e:
        print(f"Unexpected error for {eid}: {e}")

for i, pub in enumerate(publications, start=1):
    print_publication(pub, i)

File: crawler.js
const fs = require("node:fs");
const path = require("node:path");

function extractEids(searchJson) {
  const searchResults = searchJson?.["search-results"] ?? {};
  let entries = searchResults.entry ?? [];

  if (!Array.isArray(entries) && entries && typeof entries === "object") {
    entries = [entries];
  }

  return entries.filter((entry) => entry?.eid).map((entry) => entry.eid);
}

async function fetchPublication(eid, headers) {
  const url = `https://api.elsevier.com/content/abstract/eid/${encodeURIComponent(eid)}?view=FULL`;

  const response = await fetch(url, {
    method: "GET",
    headers,
    signal: AbortSignal.timeout(30000),
  });

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(`HTTP ${response.status}: ${errorText}`);
  }

  return response.json();
}

function ensureList(value) {
  if (value == null) {
    return [];
  }
  return Array.isArray(value) ? value : [value];
}

function parseAuthorName(author) {
  if (!author || typeof author !== "object") {
    return null;
  }

  const indexedName = author["ce:indexed-name"];
  if (indexedName) {
    return indexedName;
  }

  const givenName = author["ce:given-name"];
  const surname = author["ce:surname"];
  const initials = author["ce:initials"];

  if (givenName && surname) {
    return `${givenName} ${surname}`.trim();
  }
  if (surname && initials) {
    return `${surname}, ${initials}`.trim();
  }
  if (surname) {
    return surname;
  }

  return null;
}

function extractAuthors(response) {
  const authorsBlock = response?.authors ?? {};
  const authorEntries = ensureList(authorsBlock.author);

  return authorEntries.map(parseAuthorName).filter(Boolean);
}

function extractPublicationFields(pubData) {
  const response = pubData?.["abstracts-retrieval-response"] ?? {};
  const core = response.coredata ?? {};

  const allAuthors = extractAuthors(response);
  const mainAuthor = allAuthors.length > 0 ? allAuthors[0] : core["dc:creator"];
  const coAuthors = allAuthors.length > 1 ? allAuthors.slice(1) : [];

  return {
    title: core["dc:title"] ?? null,
    date_of_publication: core["prism:coverDate"] ?? null,
    publication_type: core.subtypeDescription ?? null,
    journal: core["prism:publicationName"] ?? null,
    publisher: core["dc:publisher"] ?? null,
    volume: core["prism:volume"] ?? null,
    issue: core["prism:issueIdentifier"] ?? null,
    pages: core["prism:pageRange"] ?? null,
    doi: core["prism:doi"] ?? null,
    citation_count: core["citedby-count"] ?? null,
    main_author: mainAuthor ?? null,
    co_authors: coAuthors,
  };
}

function printPublication(pub, index) {
  console.log(`\nPublication #${index}`);
  console.log("-".repeat(60));
  console.log(`Title: ${pub.title}`);
  console.log(`Date of Publication: ${pub.date_of_publication}`);
  console.log(`Type: ${pub.publication_type}`);
  console.log(`Journal: ${pub.journal}`);
  console.log(`Publisher: ${pub.publisher}`);
  console.log(`Volume: ${pub.volume}`);
  console.log(`Issue: ${pub.issue}`);
  console.log(`Pages: ${pub.pages}`);
  console.log(`DOI: ${pub.doi}`);
  console.log(`Citation Count: ${pub.citation_count}`);
  console.log(`Main Author: ${pub.main_author}`);
  console.log(`Co-Authors: ${pub.co_authors.length ? pub.co_authors.join(", ") : "None"}`);
}

function loadConfig() {
  const configPath = path.join(process.cwd(), "config.json");
  const configRaw = fs.readFileSync(configPath, "utf8");
  return JSON.parse(configRaw);
}

async function main() {
  const config = loadConfig();
  const args = process.argv.slice(2);
  const jsonOutput = args.includes("--json");
  const authorId = args.find((arg) => !arg.startsWith("--"));

  const apikey = config.apikey;
  const insttoken = config.insttoken;

  if (!apikey) {
    throw new Error("Missing 'apikey' in config.json");
  }

  if (!authorId) {
    throw new Error("Missing author ID. Usage: node crawler.js <authorId> [--json]");
  }

  const url = "https://api.elsevier.com/content/search/scopus";
  const headers = {
    Accept: "application/json",
    "X-ELS-APIKey": apikey,
  };

  if (insttoken) {
    headers["X-ELS-Insttoken"] = insttoken;
  }

  const pageSize = 200;
  let allEids = [];
  let start = 0;
  let totalResults = null;
  const failedEids = [];

  if (!jsonOutput) {
    console.log(`Fetching all publications for author ID: ${authorId}`);
  }

  while (totalResults === null || start < totalResults) {
    const params = new URLSearchParams({
      view: "STANDARD",
      query: `AU-ID(${authorId})`,
      count: String(pageSize),
      start: String(start),
    });

    const response = await fetch(`${url}?${params.toString()}`, {
      method: "GET",
      headers,
      signal: AbortSignal.timeout(30000),
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(`Search request failed: HTTP ${response.status}: ${errorText}`);
    }

    const searchResponse = await response.json();
    const pageEids = extractEids(searchResponse);

    if (totalResults === null) {
      totalResults = parseInt(searchResponse?.["search-results"]?.["opensearch:totalResults"] ?? "0", 10);
      if (!jsonOutput) {
        console.log(`Total results found: ${totalResults}`);
      }
    }

    if (!pageEids.length) {
      break;
    }

    allEids = allEids.concat(pageEids);
    if (!jsonOutput) {
      console.log(`Fetched ${pageEids.length} EIDs (total so far: ${allEids.length}/${totalResults})`);
    }

    start += pageSize;
  }

  if (!allEids.length) {
    if (!jsonOutput) {
      console.log("No EIDs found.");
    } else {
      console.log(JSON.stringify({ authorId, totalResults: 0, publications: [], failedEids: [] }));
    }
    return;
  }

  const publications = [];

  for (const eid of allEids) {
    try {
      const pubData = await fetchPublication(eid, headers);
      const limited = extractPublicationFields(pubData);
      publications.push(limited);
    } catch (error) {
      failedEids.push(eid);
      if (!jsonOutput) {
        console.error(`Failed for ${eid}: ${error.message}`);
      }
    }
  }

  if (jsonOutput) {
    console.log(
      JSON.stringify({
        authorId,
        totalResults: totalResults ?? publications.length,
        fetchedEids: allEids.length,
        publications,
        failedEids,
      })
    );
    return;
  }

  publications.forEach((pub, index) => {
    printPublication(pub, index + 1);
  });
}

main().catch((error) => {
  console.error("Crawler failed:", error.message);
  process.exit(1);
});
